eclipse-scout-core/src/dataobject/dataObjects.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2010, 2024 BSI Business Systems Integration AG
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+import {arrays, DoEntity, objects, scout} from '..';
+
+export const dataObjects = {
+  /**
+   * @returns the DO entity contribution for the given contribution class or type.
+   */
+  getContribution<TContributionDo extends DoEntity>(contributionClassOrType: DoContributionClassOrType<TContributionDo>, doEntity: DoEntityWithContributions): TContributionDo {
+    if (!doEntity._contributions?.length) {
+      return null;
+    }
+    return doEntity._contributions.find(contrib => predicate(contrib, contributionClassOrType)) as TContributionDo;
+  },
+
+  /**
+   * Adds a new DO entity contribution to the given DO entity.
+   * Existing contributions for the same contribution class are replaced. If the contribution is a plain object, existing contributions with the same type are replaced.
+   */
+  addContribution(contribution: DoEntity, doEntity: DoEntityWithContributions) {
+    if (objects.isPOJO(contribution)) {
+      scout.assertProperty(contribution, '_type');
+      dataObjects.removeContribution(contribution._type, doEntity);
+    } else {
+      // @ts-expect-error
+      dataObjects.removeContribution(contribution.constructor, doEntity);
+    }
+    doEntity._contributions = arrays.ensure(doEntity._contributions);
+    doEntity._contributions.push(contribution);
+  },
+
+  /**
+   * Removes the DO entity contributions whose class or type matches the given contribution class.
+   */
+  removeContribution<TContributionDo extends DoEntity>(contributionClassOrType: DoContributionClassOrType<TContributionDo>, doEntity: DoEntityWithContributions) {
+    arrays.removeByPredicate(doEntity._contributions, contrib => predicate(contrib, contributionClassOrType));
+    if (doEntity._contributions?.length === 0) {
+      delete doEntity._contributions;
+    }
+  }
+};
+
+function predicate(contribution: DoEntity, contributionClassOrType: DoContributionClassOrType<DoEntity>): boolean {
+  if (typeof contributionClassOrType === 'string') {
```

Comment:
this way the if in this predicate is executed for every element in the array, even though it is already known which condition must be used. Maybe extract this so that only once the condition type must be checked?

Reply:
I don't think it is worth it because there won't be many contributions